Add XIP-55: Passkey identity (#91)
xmtp/XIPs: XIPs/xip-55-passkey-identity.md

---
xip: 55
title: Passkey identity
description: Proposes a flexible identity model where new identity types can be adopted without needing coordination between all clients.

> authors: Dakota Brink (@codabrink), Nick Molnar (@neekolas), Naomi Plasterer (@nplasterer)

Author: Missing any authors here?

#### Installation IDs

XMTP address verification in the inbox ID generation algorithm will need to be tweaked to account for the new p256 address type.

## Rationale and alternatives

### Why does XMTP need Passkeys?

To authenticate messages and group membership, XMTP needs all clients to have a consistent way of validating that the installation that signed the message is a member of an inbox. This is how XMTP prevents impersonation on the network. If this were implemented without passkeys, and each new association was made equivalent to EVM addresses today, each additional account type would be a backward-incompatible change that would need to be coordinated across the network. Passkeys allow XMTP to only add one new top-level account used for authenticating messages.

Passkeys were specifically introduced to allow for inboxes that don't have any Ethereum account associated. If the goal were to only allow for additional association types *on top* of a required Ethereum address, the "Additional association types" part of this proposal could be implemented without the "New account type" section.

## Backward compatibility

> 

Author: Another other backward compatibility info we want to provide -- beyond the LibXMTP compatibility info below?

#### Phase 2

Phase 2 will add a Solana `MemberIdentifier`, and will *not* break compatibility with Phase 1 or newer LibXMTP releases due to new `MemberIdentifiers` not being able to sign identity updates, and the graceful handling of new unknown `MemberIdentifiers`.

### Inbox state

Inbox state has both a `recovery_address: String` field and an `addresses: Vec<String>` field. Because XMTP will now be supporting more chains than just Ethereum, this needs to change. First, to draw attention to the change, the fields will be renamed to `recovery_identifier` and `identifiers`, and the `String` type will be changed to a new struct: `PublicIdentifier` which is basically a more public-facing version of `MemberIdentifier`. The updated `InboxState` will look like this:

```rust
#[napi(object)]
pub struct InboxState { 
  pub inbox_id: String,
  pub recovery_identifier: PublicIdentifier,
  pub installations: Vec<Installation>,
  pub identifiers: Vec<PublicIdentifier>,
}

#[napi(object)]
pub enum PublicIdentifier {
  Ethereum(String),
  Passkey(Vec<u8>),
}

impl PublicIdentifier {
  fn display(&self) -> String {
    // ...
  }
}
```

## Reference implementation

> 

Author: Once available, provide access to reference implementation

### High-level technical implementation

#### Phase 1: Adding passkey support (breaking)

- Add a new `MemberIdentifier::Passkey` variant
  - The `Passkey` variant needs to record the relying party identifier along with the public key.
  - Rename `MemberIdentifier::Address` to `MemberIdentifier::EthAddress` to distinguish from the new `MemberIdentifier` variants coming in
- Add a new `UnverifiedSignature::Passkey` variant
  - Implement the associated `to_verified` branch logic
- Add a new `SignatureKind::P256` variant
- Rename `associate_wallet` to `associate_member` and change it to take in a `MemberIdentifier` enum instead of a `new_wallet_address`, and block on the `MemberIdentifier::Installation` variant.
- Adapt [generate_inbox_id](https://github.com/xmtp/libxmtp/blob/71b47a20020ecece139cdc0d56ab356d3f6637f1/xmtp_id/src/associations/hashes.rs#L22) to work with passkey addresses, mainly updating the valid address check.

> - Create an `Identifier` type, which is a public facing version of `MemberIdentifier` and doesn’t contain potentially sensitive metadata like the passkey’s relying partner.

Author: ```suggestion
- Create a `PublicIdentifier` type, which is a public-facing version of `MemberIdentifier` and doesn’t contain potentially sensitive metadata like the passkey’s relying partner.
```

- Rename `InboxState.account_addresses` to `InboxState.identifiers` and change the type to `Vec<Identifier>`
  - Also rename `InboxState.recovery_address` to `recovery_identifier` and change the type to `Identifier`

> 

Author: ```suggestion
- Filter out unknown `MemberIdentifiers` that may come from new associations.
```

> - Filter out unknown `MemberIdentifiers` that may come from new associations.

Author: ```suggestion
- Filter out unknown `MemberIdentifiers` that may come from new associations.

```

#### Phase 2: Adding Solana support (non-breaking)

> - Add a new `MemberIdentifier::Solana` variant

Author: ```suggestion
- Add a new `MemberIdentifier::Solana` variant that has restricted permissions. It will not be able to perform any identity updates, nor will it be allowed as the recovery identity.
```

- XMTP will **not** be adding Solana variants for `UnverifiedSignature` or `SignatureKind`

> This means that the `MemberIdentifier` in the `AddAssociation` proto may come through as `None` in old versions that haven't received the Solana update. It must be verified that this behavior is acceptable.

Reviewer: We should mention this in Phase 1 as well, since protection for this will need to be implemented there. In phase 1 can we add

```
- Filter out unknown MemberIdentifiers that may come from new associations.
```

## Security considerations

> 

Author: Any other security considerations beyond the threat model info?
